serif_swing/src/main/kotlin/serif_swing/App.kt

```diff
@@ -208,20 +255,97 @@ class SwingChatRoom(val transition: (MatrixState, Boolean) -> Unit, val panel: J
                     } else {
                         JLabel(og_image_icon)
                     }
-                } else {
-                    val message = JTextArea(message)
+                }
+                is SharedUiAudioMessage -> {
+                    val audio_url = msg.url
+                    val play_btn = JButton("Play/Pause $audio_url")
+                    play_btn.addActionListener({
+                        AudioPlayer.loadAudio(audio_url)
+                        AudioPlayer.play()
+                    })
+                    play_btn
+                }
+                else -> {
+                    val message =
+                    if(m.edits.contains(msg.id)) {
+                        val edited = m.edits.get(msg.id)!!
+                        JTextArea("${edited.message} (edited)")
+                    } else {
+                        JTextArea("$message")
+                    }
                     message.setEditable(false)
                     message.lineWrap = true
                     message.wrapStyleWord = true
                     message
                 }
+            }
+            val msg_action_popup = JPopupMenu()
+
+            val reply_option = JMenuItem("Reply")
+            reply_option.addActionListener({
+                println("Now writing a reply")
+                replied_event_id = msg.id
+            })
+            val edit_option = JMenuItem("Edit")
+            edit_option.addActionListener({
+                println("Now editing a message")
+                edited_event_id = msg.id
+                message_field.text = msg.message
+            })
+            val show_src_option = JMenuItem("Show Source")
+            show_src_option.addActionListener({
+                val json_str = m.getEventSrc(msg.id)
+
+                val window = SwingUtilities.getWindowAncestor(panel)
+                val dim = window.getSize()
+                val h = dim.height
+                val w = dim.width
+                val dialog = JDialog(window, "Event Source")
+
+                val top_panel = JPanel()
```

Comment:
I think this panel doesn't do anything, is it necessary?

Reply:
Good call, we can indeed get rid of it